File: 20_ValidParentheses/20.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
  stack = []
  pairs = {
    ")" => "(",
    "]" => "[",
    "}" => "{"
  }
  s.each_char { |char|
    if "({[".include? char
      stack.push char
    else
      if stack.empty? then return false end
      if not stack.pop == pairs[char] then return false end
    end
  }
  return stack.size == 0
end

File: src/20_ValidParentheses/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        pair = {
            ")":"(",
            "]":"[",
            "}":"{",
        }
        for char in s:
            if char in "({[":
                stack.append(char)
            else:
                try:
                    c = stack.pop()
                    if c != pair[char]:
                        return False
                    else:
                            return false
                except:
                    return False
        
        return not len(stack)

  